File: exptr_frontend/src/types/operation.ts
// =============================================================================
// Operation Types
// =============================================================================

import type { ApiResponse } from "./common";

// -----------------------------------------------------------------------------
// Enums / Literal Types
// -----------------------------------------------------------------------------

export type OperationType = "income" | "expense";
export type Currency = "USD" | "EUR" | "GBP" | "RUB" | string;

// -----------------------------------------------------------------------------
// Entity Types
// -----------------------------------------------------------------------------

export interface Operation {
  id: string;
  name: string;
  type: OperationType;
  amount: number;
  currency: Currency;
  category_id: string;
  comment?: string;
  user_id: string;
  created_at: string;
  updated_at: string;
}

// -----------------------------------------------------------------------------
// Request Types
// -----------------------------------------------------------------------------

export interface OperationRequest {
  name: string;
  type: OperationType;
  amount: number;
  currency: Currency;
  category_id: string;
  comment?: string;
  user_id: string;
  created_at: string;
  updated_at?: string;
}

// -----------------------------------------------------------------------------
// Response Types
// -----------------------------------------------------------------------------

export type CreateOperationResponse = ApiResponse;

export type UpdateOperationResponse = ApiResponse;

export interface GetOperationsResponse extends ApiResponse {
  operations?: Operation[];
}

File: exptr_api/internal/domain/operation.go
package domain

import (
	"github.com/google/uuid"
)

type Operation struct {
	BaseEntity
	UserID     uuid.UUID `json:"user_id" gorm:"type:uuid;index"`
	CategoryID uuid.UUID `json:"category_id" gorm:"type:uuid;not null;index"`
	Amount     int       `json:"amount" gorm:"type:decimal(19,4);not null"`
	Currency   string    `json:"currency" gorm:"type:varchar(10);not null"`
	Name       string    `json:"name" gorm:"type:varchar(255);not null"`
	Comment    string    `json:"comment" gorm:"type:text"`
	Type       string    `json:"type" gorm:"type:varchar(255)"`
}

func (Operation) TableName() string {
	return "operations"
}
